File: src/com/company/Parser.java
package com.company;

import java.util.ArrayList;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Created by LaoFan on 3/29/19.
 */
public class Parser {
    private Node headNode;
    private int netLength;
    private String netCode;
    private String name;
    private ArrayList<Token> tokenList;
    private Node currentNode; // maybe not need
    // indicate the current scope, should be parameter
//    private int leftIndex;
//    private int rightIndex;

    public static Tree parse(String code) {
        Parser p = new Parser(code);
        return new Tree(p.name, p.headNode);
    }

    private static String removeBlank(String code) {
        String dest = null;
        if (code != null) {
            Pattern p = Pattern.compile("\\s*|\t|\r|\n"); // TODO check if it remove the space, \s is space
            Matcher m = p.matcher(code);
            dest = m.replaceAll("");
        }
        return dest;
    }

    private Parser(String code) {
        netCode = Parser.removeBlank(code);
        netLength = netCode.length();
        headNode = tokenize().check().doParse(3);
    }

    private Parser tokenize() {
        tokenList = new ArrayList<Token>();
        StringBuilder currentBuffer = new StringBuilder();
        int currentLevel = 0;

        for (int i = 0; i < netLength; ++i) {
            switch (netCode.charAt(i)) {
                case '(':
                    ++currentLevel;
                    tokenList.add(new Token("(", currentLevel, TokenType.leftBracket));
                    break;
                case ')':
                    tokenList.add(new Token(")", currentLevel, TokenType.rightBracket));
                    --currentLevel;
                    break;
                case '[':
                    ++currentLevel;
                    tokenList.add(new Token("[", currentLevel, TokenType.leftSquareBracket));
                    break;
                case ']':
                    tokenList.add(new Token("]", currentLevel, TokenType.rightSquareBracket));
                    --currentLevel;
                    break;
                case ' ':
                    tokenList.add(new Token(currentBuffer.toString(), currentLevel, TokenType.word));
                    currentBuffer.delete(0, currentBuffer.length());
                    break;
                default:
                    char c = netCode.charAt(i);
                    if (!Character.isLetterOrDigit(c)) {
                        throw new IllegalArgumentException("Code has grammar error: " + currentBuffer);
                    } else if (currentBuffer.length() != 0){
                        currentBuffer.append(c);
                    }
            }
            if (currentLevel < 0) {
                throw new IllegalArgumentException("Bracket is not pair: " + currentBuffer);
            }
        }

        if (currentLevel > 0) {
            throw new IllegalArgumentException("Still have bracket is not paired");
        }
        return this;
    }

    private Parser check() {
        // Process define
        if (tokenList.get(0).content() == "("
                && tokenList.get(1).content().toLowerCase() == "define"
                && tokenList.get(2).type() == TokenType.word) {
            name = tokenList.get(2).content();
            if (tokenList.get(3).type() != TokenType.leftBracket
                    && tokenList.get(3).type() != TokenType.leftSquareBracket) {
                throw new IllegalArgumentException("After define name, it should be '(' or '[' there");
            }
        }
        // TODO check detail about pair is [], not like [)

        // check word place
        TokenType lastType = null, currentType = null, nextType = tokenList.get(3).type();
        for (int i = 3; i < tokenList.size(); ++i) {
            currentType = nextType;
            nextType = tokenList.get(i + 1).type();
            if (currentType == TokenType.word) {
                if (lastType != TokenType.leftBracket || nextType != TokenType.rightBracket) {
                    throw new IllegalArgumentException("The word can only be allowed between ( )");
                }
            }
            lastType = currentType;
        }
        return this;
    }

    private Node doParse(int i) {
        int pairedOne;
        switch (tokenList.get(i).type()) {
            case leftBracket:
                pairedOne = findPair(i, TokenType.rightBracket);
                return processBracket(i, pairedOne);
            case leftSquareBracket:
                pairedOne = findPair(i, TokenType.rightSquareBracket);
                return processSquareBracket(i, pairedOne);
            default:
                // TODO should declare NAME in the document
                throw new IllegalArgumentException("Found the illegal element after the NAME");
        }

    }

    private int findPair(int i, TokenType destType) {
        int srcLevel = tokenList.get(i).level();
        for (; i < tokenList.size(); ++i) {
            if (tokenList.get(i).type() == destType && tokenList.get(i).level() == srcLevel) {
                return i;
            }
        }
        throw new IllegalArgumentException("Not found the paired one of " + tokenList.get(i).content());
    }

    private int findPair(int i) {
        switch (tokenList.get(i).type()) {
            case leftBracket:
                return findPair(i, TokenType.rightBracket);
            case leftSquareBracket:
                return findPair(i, TokenType.rightSquareBracket);
            default:
                throw new IllegalArgumentException("Not found the paired one of " + tokenList.get(i).content());
        }
    }

    private Node processBracket(int begin, int end) {
        switch (end - begin) {
            case 1: // () should in grammar check? TODO
                throw new IllegalArgumentException("The content between two paired bracket should not be empty");
            case 2: // (word)
                return new Node(tokenList.get(begin + 1).content());
            default: // like (()[])
                Node last;
                Node first = last = doParse(begin + 1);
                begin = findPair(begin + 1);

                while (begin < end) {
                    // there are some info lose between the two function
                    Node n = doParse(begin + 1);
                    last.addNode(n);
                    // update
                    last = n;
                    begin = findPair(begin + 1); // actually begin point to the last pair's end
                }
                return first; // or should return last
        }
    }

    private Node processSquareBracket(int begin, int end) {
        switch (end - begin) {
            case 1: // [] should in grammar check? TODO
                throw new IllegalArgumentException("The content between two paired bracket should not be empty");
            case 2: // [word]
                return new Node(tokenList.get(begin + 1).content());
            default: // like [()[]]
                Node n = new Node("");
                while (begin < end) {
                    n.addNode(doParse(begin + 1));
                    begin = findPair(begin + 1); // actually begin point to the last pair's end
                }
                return n;
        }
    }
}

File: CSharpSrc/Parser.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace ConfigureParser
{
    public class Parser
    {
        private readonly Node _headNode;
        private string _name;

        private readonly string _netCode;
        private List<Token> _tokenList;
        private readonly List<Node> _nodes = new List<Node>(); // Convenient to equip operation, so just record the Not "" Node

        public static Tree Parse(string code)
        {
            var p = new Parser(code);
            return new Tree(p._name, p._headNode, p._nodes);
        }

        private Parser(string code)
        {
            _netCode = Parser.ReplaceLine(code);
            _headNode = Tokenize().Check().DoParseFrom(3, FindPair(3)).Head; // 3 is the first element to parse
        }

        private Parser Tokenize()
        {
            _tokenList = new List<Token>();
            var currentBuffer = new StringBuilder();
            var currentLevel = 0;

            foreach (var c in _netCode)
            {
                switch (c)
                {
                    case '(':
                        ++currentLevel;
                        _tokenList.Add(new Token("(", currentLevel, TokenType.LeftBracket));
                        break;
                    case ')':
                        if (currentBuffer.Length > 0) // (word)
                        {
                            _tokenList.Add(new Token(currentBuffer.ToString(), currentLevel, TokenType.Word));
                            currentBuffer.Clear();
                        }
                        _tokenList.Add(new Token(")", currentLevel, TokenType.RightBracket));
                        --currentLevel;
                        break;
                    case '[':
                        ++currentLevel;
                        _tokenList.Add(new Token("[", currentLevel, TokenType.LeftSquareBracket));
                        break;
                    case ']':
                        _tokenList.Add(new Token("]", currentLevel, TokenType.RightSquareBracket));
                        --currentLevel;
                        break;
                    case ' ': // TODO next time not use space to divide token
                        if (currentBuffer.Length > 0) // "  " two spaces
                        {
                            _tokenList.Add(new Token(currentBuffer.ToString(), currentLevel, TokenType.Word));
                            currentBuffer.Clear();
                        }
                        break;
                    default:
                        if (char.IsLetterOrDigit(c) || c == '.')
                        {
                            currentBuffer.Append(c);
                        }
                        else
                        {
                            throw new ArgumentException("Code has illegal char: " + currentBuffer);
                        }
                        break;
                }
                if (currentLevel < 0)
                {
                    throw new ArgumentException("Bracket is not pair: " + currentBuffer);
                }
            }

            if (currentLevel > 0)
            {
                throw new ArgumentException("Still have bracket is not paired");
            }
            return this;
        }

        private Parser Check()
        {
            // Process define
            if (_tokenList[0].Content == "("
                && _tokenList[1].Content.ToLower() == "define"
                && _tokenList[2].Type == TokenType.Word)
            {
                _name = _tokenList[2].Content;
                if (_tokenList[3].Type != TokenType.LeftBracket
                    && _tokenList[3].Type != TokenType.LeftSquareBracket)
                {
                    throw new ArgumentException("After define name, it should be '(' or '[' there");
                }
            }
            // TODO check detail about pair is [], not like [)

            // check word place
            var lastType = _tokenList[1].Type;
            var nextType = _tokenList[3].Type;
            for (var i = 3; i < _tokenList.Count - 1; ++i) // - 1 maybe occur some problem, some other grammar is not check, like () abc
            {
                var currentType = nextType;
                nextType = _tokenList[i + 1].Type;
                if (currentType == TokenType.Word)
                {
                    if (lastType != TokenType.LeftBracket || nextType != TokenType.RightBracket)
                    {
                        throw new ArgumentException("The word can only be allowed between ( )");
                    }
                }
                lastType = currentType;
            }
            return this;
        }

        /// <returns>The pair of head and tail of a linked list</returns>
        private HeadTail DoParseFrom(int start, int end)
        {
            switch (_tokenList[start].Type)
            {
                case TokenType.LeftBracket:
                    return ProcessBracket(start, end);
                case TokenType.LeftSquareBracket:
                    return ProcessSquareBracket(start, end);
                default:
                    // TODO should declare NAME in the document
                    throw new ArgumentException("Illegal character was found after the NAME");
            }
        }

        /// <returns>The pair of head and tail of a linked list</returns>
        private HeadTail ProcessBracket(int begin, int end)
        {
            switch (end - begin)
            {
                case 1: // () should in grammar check? TODO
                    throw new ArgumentException("The content between two paired bracket should not be empty");
                case 2: // (word)
                    return new HeadTail(MakeNewNode(_tokenList[begin + 1].Content));
                default: // like (()[])
                    var e = FindPair(begin + 1);
                    var first = DoParseFrom(begin + 1, e);
                    var tail = first.Tail;
                    begin = e;

                    while (begin < end - 1)
                    {
                        begin = begin + 1;
                        e = FindPair(begin);
                        var current = DoParseFrom(begin, e);

                        tail.AddChild(current.Head);
                        tail = current.Tail;
                        // update
                        begin = e;
                    }

                    return new HeadTail(first.Head, tail);
            }
        }

        /// <returns>The pair of head and tail of a linked list</returns>
        private HeadTail ProcessSquareBracket(int begin, int end)
        {
            switch (end - begin)
            {
                case 1: // [] should in grammar check? TODO
                    throw new ArgumentException("The content between two paired bracket should not be empty");
                case 2: // [word]
                    throw new ArgumentException("The content between two paired bracket should not be just one word directly");
                default: // like [()[]]
                    var head = MakeNewNode("");
                    var tail = MakeNewNode("");
                    
                    while (begin < end - 1) // [()]
                    {
                        begin = begin + 1;
                        var e = FindPair(begin);
                        var current = DoParseFrom(begin, e);
                        head.AddChild(current.Head);
                        current.Tail.AddChild(tail);

                        // update
                        begin = e;
                    }
                    return new HeadTail(head, tail);
            }
        }

        private int FindPair(int i, TokenType destType)
        {
            var srcLevel = _tokenList[i].Level;
            for (; i < _tokenList.Count; ++i)
            {
                if (_tokenList[i].Type == destType && _tokenList[i].Level == srcLevel)
                {
                    return i;
                }
            }
            throw new ArgumentException("Not found the paired one of " + _tokenList[i].Content);
        }

        private int FindPair(int i)
        {
            switch (_tokenList[i].Type)
            {
                case TokenType.LeftBracket:
                    return FindPair(i, TokenType.RightBracket);
                case TokenType.LeftSquareBracket:
                    return FindPair(i, TokenType.RightSquareBracket);
                default:
                    throw new ArgumentException("Not found the paired one of " + _tokenList[i].Content);
            }
        }

        private Node MakeNewNode(string content)
        {
            // use Linq to check if this content is existed in _node
            var n = new Node(content);
            if (content != "")
            {
                _nodes.Add(n);
            }
            return n;
        }

        private static string ReplaceLine(string code)
        {
            return code
                .Replace("\n", " ")
                .Replace("\t", " ")
                .Replace("\r", " ");
        }

        /// <summary>
        /// This class is to help code to be readable
        /// </summary>
        private class HeadTail
        {
            public Node Head { get; }
            public Node Tail { get; }

            public HeadTail(Node head, Node tail)
            {
                Head = head;
                Tail = tail;
            }

            public HeadTail(Node onlyOne)
            {
                Head = Tail = onlyOne;
            }
        }

    }
}
